### PR TITLE
[SPARK-33615][K8S] Make 'spark.archives' working in Kubernates

### PR DESCRIPTION
**Pull request by HyukjinKwon**
### What changes were proposed in this pull request?

This PR proposes to make `spark.archives` configuration working in Kubernates.
It works without a problem in standalone cluster but there seems a bug in Kubernates.
It fails to fetch the file on the driver side as below:

```
20/12/03 13:33:53 INFO SparkContext: Added JAR file:/tmp/spark-75004286-c83a-4369-b624-14c5d2d2a748/spark-examples_2.12-3.1.0-SNAPSHOT.jar at spark://spark-test-app-48ae737628cee6f8-driver-svc.spark-integration-test.svc:7078/jars/spark-examples_2.12-3.1.0-SNAPSHOT.jar with timestamp 1607002432558
20/12/03 13:33:53 INFO SparkContext: Added archive file:///tmp/tmp4542734800151332666.txt.tar.gz#test_tar_gz at spark://spark-test-app-48ae737628cee6f8-driver-svc.spark-integration-test.svc:7078/files/tmp4542734800151332666.txt.tar.gz with timestamp 1607002432558
20/12/03 13:33:53 INFO TransportClientFactory: Successfully created connection to spark-test-app-48ae737628cee6f8-driver-svc.spark-integration-test.svc/172.17.0.4:7078 after 83 ms (47 ms spent in bootstraps)
20/12/03 13:33:53 INFO Utils: Fetching spark://spark-test-app-48ae737628cee6f8-driver-svc.spark-integration-test.svc:7078/files/tmp4542734800151332666.txt.tar.gz to /tmp/spark-66573e24-27a3-427c-99f4-36f06d9e9cd5/fetchFileTemp2665785666227461849.tmp
20/12/03 13:33:53 ERROR SparkContext: Error initializing SparkContext.
java.lang.RuntimeException: Stream '/files/tmp4542734800151332666.txt.tar.gz' was not found.
	at org.apache.spark.network.client.TransportResponseHandler.handle(TransportResponseHandler.java:242)
	at org.apache.spark.network.server.TransportChannelHandler.channelRead0(TransportChannelHandler.java:142)
	at org.apache.spark.network.server.TransportChannelHandler.channelRead0(TransportChannelHandler.java:53)
```

This is because `spark.archives` was not actually added on the driver side correctly. The changes here fix it by adding and resolving URIs correctly.

### Why are the changes needed?

`spark.archives` feature can be leveraged for many things such as Conda support. We should make it working in Kubernates as well.
This is a bug fix too.

### Does this PR introduce _any_ user-facing change?

No, this feature is not out yet.

### How was this patch tested?

I manually tested with Minikube 1.15.1. For an environment issue (?), I had to use a custom namespace, service account and roles. `default` service account does not work for me and complains it doesn't have permissions to get/list pods, etc.

```bash
minikube delete
minikube start --cpus 12 --memory 16384
kubectl create namespace spark-integration-test
cat <<EOF | kubectl apply -f -
apiVersion: v1
kind: ServiceAccount
metadata:
  name: spark
  namespace: spark-integration-test
EOF
kubectl create clusterrolebinding spark-role --clusterrole=edit --serviceaccount=spark-integration-test:spark --namespace=spark-integration-test
dev/make-distribution.sh --pip --tgz -Pkubernetes
resource-managers/kubernetes/integration-tests/dev/dev-run-integration-tests.sh --spark-tgz `pwd`/spark-3.1.0-SNAPSHOT-bin-3.2.0.tgz  --service-account spark --namespace spark-integration-test
```

### CODE REVIEW

**Comment by dongjoon-hyun** [comment]
Yay! Great!

**Comment by dongjoon-hyun** [comment]
Are we waiting Jenkins K8s IT?

**Comment by HyukjinKwon** [comment]
Oh yeah. and I am also testing in my local at the same time.

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by HyukjinKwon** [comment]
I managed to set up my local env but there seems a bug in Kubernates - the files does not exist in driver side although it's client mode, and the tar.gz file is correctly created. I am debugging it.

**Comment by HyukjinKwon** [comment]
cc @tgravescs FYI. I am testing K8S here for `spark.archives`.

**Comment by tgravescs** [comment]
great thanks, I'll keep an eye on this and review when you are ready.  I have a local k8s setup as well if you need particular help to validate something

File: resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala
Comment (HyukjinKwon) [inline resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala:1632]: This is also unrelated changes but I just piggybacked here as it looks pretty clear to me.

**Comment by HyukjinKwon** [comment]
@dongjoon-hyun and @tgravescs, this PR is ready for a look. At least the tests pass. I looked up related commits in the history too. I am pretty sure it's almost ready to go.

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/DepsTestsSuite.scala
Comment (HyukjinKwon) [inline resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/DepsTestsSuite.scala:166]:
```suggestion
  test("SPARK-33615: Launcher client archives", k8sTestTag, MinikubeTag) {
```

**Comment by HyukjinKwon** [comment]
BTW, for compeleteness, I tested in standalone cluster again.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM in general if CIs pass. Thank you so much!
I'll leave this to you guys, @HyukjinKwon and @tgravescs .

**Comment by HyukjinKwon** [comment]
Thank you so much @dongjoon-hyun!

**Comment by HyukjinKwon** [comment]
There's discommision test failure which I don't think is related .. but I will still take a look tomorrow for doubly sure.

**Comment by dongjoon-hyun** [comment]
The new test case passed and the other two failures are known flaky ones which is irrelevant to this. I believe you can merge this if GitHub Action passed.
```
- SPARK-33615: Launcher client archives
```

```
- Run in client mode. *** FAILED ***
  io.fabric8.kubernetes.client.KubernetesClientException: Failure executing: POST at: https://192.168.39.142:8443/api/v1/namespaces/6ccb7b90fb5c4841bae64ebdaff6d687/pods. Message: No API token found for service account "default"
...
- Test basic decommissioning *** FAILED ***
  The code passed to eventually never returned normally. Attempted 173 times over 3.013441304633333 minutes. Last failure message: "++ id -u
  " did not contain "Decommission executors" The application did not complete, driver log did not contain str Decommission executors. (KubernetesSuite.scala:402)
```

**Comment by HyukjinKwon** [comment]
The Hive test failure is happening globally so it should be unrelated to the current PR.
Please allow me to merge it I am going to cut the branch now. 

Thank you @dongjoon-hyun and @tgravescs.

@tgravescs, please let me know if there's any concern. I will address them in a followup

**Comment by HyukjinKwon** [comment]
Merged to master.

**Pull request closed without merging**
